### PR TITLE
A pull request for merging the "Exponentiation function" to the main branch

### LINKED ISSUE #66: Exponentiation_Function
Description:
Feature: Exponentiation Function

Implement a function named "exponentiation" that computes the result of raising a base number to a given exponent. This function should handle both positive and negative exponents and work with both float and integer bases.

Parameters:
1. base (float): The base number to be raised.
2. exponent (int): The exponent to which the base is raised.

Returns:
1. float: The result of the base raised to the given exponent.

Requirements:
Type Checking:
Raise a "TypeError" if the base is not a float or int.
Raise a "TypeError" if the exponent is not an int.

Functionality:
Calculate the result of base^exponent.

### PR DESCRIPTION
**Pull request by zeozeph**
…ittests

---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [ ] The function has correct type annotations
- [ ] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Ridwanayinde — COMMENTED** [review]
@zeozeph I will take on the review of your code

File: solutions/exponentiation.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      2.25
+      >>> exponentiation(2, -2)
+      0.25
+    """
+    if not isinstance(base, (float, int)):
+        raise TypeError("The base must be a float or int")
+    if not isinstance(exponent, int):
+        raise TypeError("The exponent must be an int")
+
+    return base**exponent
```
Comment (Ridwanayinde) [inline solutions/exponentiation.py:46]: @zeozeph Since you return float in your code, to ensure consistency, kindly input add float after the return. return float (base**exponent)

File: solutions/tests/test_exponentiation.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    def test_type_error_for_invalid_base(self):
+        """It should raise a TypeError for an invalid base"""
+        with self.assertRaises(TypeError):
+            exponentiation("2", 3)
+
+    def test_type_error_for_invalid_exponent(self):
+        """It should raise a TypeError for an invalid exponent"""
+        with self.assertRaises(TypeError):
+            exponentiation(2, "3")
```
Comment (Ridwanayinde) [inline solutions/tests/test_exponentiation.py:48]: I like your test cases. kudos to you. Can you also test some large numbers to see if the exponentiation function handles very large exponentiation like (2, 100). Also, can you include the test of some edge cases like (0, 3) and (0, 0).

File: solutions/exponentiation.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      8.0
+      >>> exponentiation(5, 0)
+      1.0
+      >>> exponentiation(1.5, 2)
+      2.25
+      >>> exponentiation(2, -2)
+      0.25
+    """
+    if not isinstance(base, (float, int)):
+        raise TypeError("The base must be a float or int")
```
Comment (Ridwanayinde) [inline solutions/exponentiation.py:42]: This is good but instead of the generic type. You can include the actual value received to provide more context for the user. raise TypeError (f "Invalid base: {base}. The base must be a float or int")

File: solutions/exponentiation.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    if not isinstance(exponent, int):
+        raise TypeError("The exponent must be an int")
```
Comment (Ridwanayinde) [inline solutions/exponentiation.py:44]: The same thing applies here. raise TypeError (f " Invalid exponent :{exponet}. The exponent must be an int")

File: solutions/exponentiation.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      >>> exponentiation(5, 0)
+      1.0
+      >>> exponentiation(1.5, 2)
+      2.25
+      >>> exponentiation(2, -2)
+      0.25
+    """
+    if not isinstance(base, (float, int)):
+        raise TypeError("The base must be a float or int")
+    if not isinstance(exponent, int):
```
Comment (Ridwanayinde) [inline solutions/exponentiation.py:43]: If the values are too large and are valid but the range is out of bounds. Kindly raise an exception error. e.g 
if abs(exponent) > 1000. I hope you understand.

**Review by Ridwanayinde — COMMENTED** [review]
@zeozeph. I just reviewed your code. Kindly check it out.

File: solutions/exponentiation.py (hunk 1 above)
Comment (zeozeph) [reply solutions/exponentiation.py:46]: Thank you for this suggestion, i have a adapted it.

File: solutions/tests/test_exponentiation.py (hunk 2 above)
Comment (zeozeph) [reply solutions/tests/test_exponentiation.py:48]: Yes, my functions handles extremely large numbers. For avoidance of doubt, I have added a test case in this light.

Comment (zeozeph) [reply solutions/tests/test_exponentiation.py:48]: More so, the line of code you referenced here is meant for a "string" error test case. That is, a situation where a string is used as the exponent of a number in lieu of a valid number.

File: solutions/exponentiation.py (hunk 5 above)
Comment (zeozeph) [reply solutions/exponentiation.py:43]: The function should handle large numbers like (2, 1000) correctly. Python's built-in integers can handle arbitrarily large numbers, so we should not encounter overflow errors like you might in languages with fixed-width integers. For the sake of avoidance of doubt, I have added a test case in this light.

File: solutions/exponentiation.py (hunk 4 above)
Comment (zeozeph) [reply solutions/exponentiation.py:44]: Duly noted, thank you for your great review and suggestions

**Comment by zeozeph** [comment]
Dear @Ridwanayinde, I have applied all your recommendations. Kindly second-guess this and merge to the main branch. Thanks a lot for your great suggestions they were truly insightful.

File: solutions/tests/test_exponentiation.py (hunk 2 above)
Comment (Ridwanayinde) [reply solutions/tests/test_exponentiation.py:48]: In this case, I wanted to reference all your test lines, but when I added the comment, it only showed that line.

**Review by Ridwanayinde — CHANGES_REQUESTED** [review]
@zeozeph. Kindly check your py formatting for both pull and push request.  The code is okay now but it is not passing the CI checks. Kindly format the code using ruff format file.py and also check the line length. Is it still within the recommended line length? I mean the changes I ask to improve your code, is it still within the recommended length?
